### ISSUE TITLE
Feature: dynamic number of lines, resize and reflow based on context and while typing

### ISSUE THREAD

**Issue by ToyKeeper** [romkatv/powerlevel10k#2164]
P10K seems generally pretty awesome, but I haven't fully made the switch yet because I haven't found a way to make it do some things I really like about my janky old custom prompt.  The largest of these is the ability to automatically adapt its layout and number of lines on the fly.

It's probably easiest to convey a lot of it with a screenshot, so... here's an example:
![custom zsh prompt](http://toykeeper.net/tmp/zsh-smart-prompt.jpg)

It uses a 1-line prompt by default, unless there is too much info to fit in 1 line.  Then it switches to a 2-line prompt by moving some segments.  In my config, it moves the directory and prompt symbol to the 2nd line while keeping everything else on the 1st line.  This can happen while typing a command, and it also goes back to 1 line whenever things get short enough to fit again.  Like, press backspace enough times, and it goes back to 1 line if possible.

Although not shown in the pic, the dynamic sizing also factors in POSTDISPLAY for the auto-suggestion module I have installed.  So it reflows when the suggested completion doesn't fit.

It does not let the user (or the postdisplay) overwrite the right-side prompt.  If what they're typing is long enough to cover that up, it switches to 2-line mode.

Additionally, if the right-side prompt can't fit even in 2-line mode, like if the window is too narrow, it uses a multi-stage algorithm to try to make things fit.  First, it changes the date/time format to something shorter: it removes the date, then it removes the seconds, then it removes the timestamp entirely.  If things still can't fit, it removes right-side segments one at a time until the remainder is short enough.  The left-most right-prompt segment is assumed to be the most important, so it gets removed last.

I didn't see way to do this in p10k though, and it looked like it would be complicated to add.  So instead of a patch, I have only a feature request.

**Comment by romkatv** [comment]
You can do some of this by invoking `p10k display` from `p10k-on-pre-prompt` (invoked before each prompt) and `p10k-on-post-widget` (invoked after each zle widget). See `p10k help display` for docs.

In general, I cannot recommend trying to replicate your previous prompt exactly in p10k. If you like your prompt, keep it. If you like p10k better, switch and change your habits.

If you do decide to give p10k a shot, the config preferred by power users is something like this (although probably without the dotted line):

![prompt](https://raw.githubusercontent.com/romkatv/powerlevel10k-media/master/transient-prompt.gif)

The only thing that dynamically shrinks on the TTY resize is the current directory:

![cwd](https://raw.githubusercontent.com/romkatv/powerlevel10k-media/master/directory-truncation.gif)

**Comment by ToyKeeper** [comment]
Thanks for such a quick reply!  <3

I saw and tried the transient prompt feature before opening this feature request.  It's nice... but not my cup of tea, so to speak.  I often rely on having the prompt info in scrollback, so I don't want it to disappear after pressing Enter.

The pre-prompt and post-widget hooks seem really promising for implementing dynamic prompt layout changes, and I'll definitely have to try that.  It might resolve things for me personally... but if someone else wants similar behavior, they'd have to go through the same process as me.

So I considered trying to add a more general framework for handling it automatically, but it quickly became clear that it would require me to get intimately familiar with p10k's internals so it'd probably be better to ask someone who has a better grasp of how to do it.  Diving into internals like that as a noob is a recipe for breaking things I don't even know exist.

I'm not sure how difficult it would be in p10k, but the method I used was: There are two separate prompt templates.  I defined the 1-line template as a "west" prompt and "east" prompt... and when west + buffer + postdisplay + east is greater than $columns, it rebuilds the prompt using the second template.  The second template has northwest, northeast, southwest, and southeast prompts (but I left southeast blank).  And the northeast one has some shortening logic applied to it to try extra hard to fit as much as possible.  Oh, and a filler string to go between northwest and northeast.

It seems like this general concept could be useful in p10k too, and could be implemented in a way such that users could use it without any custom hooks.  Basically just a short and long prompt definition, which it would switch between automatically if the user enabled an option to do so.

As for shortening and omitting segments...  Perhaps segments could return an array instead of a string, containing multiple renderings from longest to shortest, and the code which combines them could pick which versions and which segments to use to fit as much info as possible.  Or perhaps segments could support a "shorten" signal to request a smaller version on demand... and if they can't go any shorter, they get omitted?  Maybe shorten-able prompts would need some sort of wrapper to mark them as normal or shorten-able?  Maybe the user should define the order things should be omitted, instead of just removing the head or tail each time?  I don't know what kind of implementation would fit best in p10k.

Any thoughts on what might be easiest within the existing code structure?
